Deepstream 5.0 Subscription Implementation

Shubhamturakhia opened the pull request:
Add subscribe implementation for client to device messaging

Shubhamturakhia committed:
Add msgapi_subscribe implementation

Shubhamturakhia committed:
Modify broker.c file

Shubhamturakhia committed:
Added subcribe implementation

Shubhamturakhia committed:
Added Subscribe implementation

lummish commented on aws_protocol_adaptor/device_client/aws_nvmsgbroker.c:
```diff
@@ -243,6 +243,44 @@ NvDsMsgApiErrorType nvds_msgapi_send_async(NvDsMsgApiHandle h_ptr, char *topic,
+
+NvDsMsgApiErrorType nvds_msgapi_subscribe(NvDsMsgApiHandle h_ptr, char **topics, int num_topics, nvds_msgapi_subscribe_request_cb_t cb, void *user_ctx)
+{
+	printf("Subscribe called\n");
```
You should use the IOT logging functions.

lummish commented on aws_protocol_adaptor/device_client/aws_nvmsgbroker.c:
```diff
@@ -243,6 +243,44 @@ NvDsMsgApiErrorType nvds_msgapi_send_async(NvDsMsgApiHandle h_ptr, char *topic,
+		return NVDS_MSGAPI_ERR;
+	}
+	IOT_INFO("Successfully subscribed");
+	g_free(client);
```
You realize this will destroy the client, right? Why would we want to do that?

lummish commented on aws_protocol_adaptor/device_client/aws_nvmsgbroker.c:
```diff
@@ -243,6 +243,44 @@ NvDsMsgApiErrorType nvds_msgapi_send_async(NvDsMsgApiHandle h_ptr, char *topic,
+
+	if ((h_ptr == NULL) || (topics == NULL) || (num_topics <= 0))
+	{
+		IOT_ERROR("Essensial args missing for function nvds_msgapi_subscribe: %d, %d, %d\n", (h_ptr == NULL), (topics == NULL), (num_topics == NULL);
```
Typo: "essensial"

Shubhamturakhia replied:
Thats the typo in the whole file I will correct that in all other places as well

Shubhamturakhia committed:
Minor changes

lummish commented on aws_protocol_adaptor/device_client/aws_nvmsgbroker.c:
```diff
@@ -243,6 +243,44 @@ NvDsMsgApiErrorType nvds_msgapi_send_async(NvDsMsgApiHandle h_ptr, char *topic,
+	}
+	IoT_Error_t rc = FAILURE;
+	AWS_IoT_Client *client = (AWS_IoT_Client *)h_ptr;
+	rc = aws_iot_mqtt_subscribe(client, topics, strlen(topics), QOS0, subscribe_callback_handler, NULL);
```
Please review [the documentation](http://aws-iot-device-sdk-embedded-c-docs.s3-website-us-east-1.amazonaws.com/aws__iot__mqtt__client__subscribe_8c.html) for this function.

There are a few problems here:
1. This likely will not compile as you are not using the correct interface for `aws_iot_mqtt_subscribe`
2. `aws_iot_mqtt_subscribe` only subscribes to a single topic (see 1).
3. You're calling `strlen` on an array of strings, not a string.
4. You're not using `cb` at all. Instead you've provided your own callback, so the argument (and by extension, the subscription) is essentially useless.
5. You're not using `user_ctx`

lummish commented on aws_protocol_adaptor/device_client/aws_nvmsgbroker.c:
```diff
@@ -243,6 +243,44 @@ NvDsMsgApiErrorType nvds_msgapi_send_async(NvDsMsgApiHandle h_ptr, char *topic,
 	return NVDS_MSGAPI_OK;
 }
 
+/* ************************************************************************* */
+// Subscribe def
+/* ************************************************************************* */
```
Get rid of this comment.

lummish commented on aws_protocol_adaptor/device_client/aws_nvmsgbroker.c:
```diff
@@ -243,6 +243,44 @@ NvDsMsgApiErrorType nvds_msgapi_send_async(NvDsMsgApiHandle h_ptr, char *topic,
+// Subscribe def
+/* ************************************************************************* */
+
+void subscribe_callback_handler(AWS_IoT_Client *pClient, char *topicName, uint16_t topicNameLen,
+								IoT_Publish_Message_Params *params, void *pData)
+{
+	IOT_INFO("Subscribe callback");
+	IOT_INFO("%.*s\t%.*s", topicNameLen, topicName, (int)params->payloadLen, (char *)params->payload);
+}
```
Why is this here? See comments below.

lummish reviewed (changes requested):
Major issues.

Shubhamturakhia committed:
Changes according to PR review

lummish commented on aws_protocol_adaptor/device_client/aws_nvmsgbroker.c:
```diff
@@ -243,41 +243,32 @@ NvDsMsgApiErrorType nvds_msgapi_send_async(NvDsMsgApiHandle h_ptr, char *topic,
 	{
-		IOT_ERROR("Unable to subscribe, error: %d\n", rc);
-		return NVDS_MSGAPI_ERR;
+		rc = aws_iot_mqtt_subscribe(client, topics[i], strlen(topics[i]), QOS0, cb, user_ctx);
```
This will not compile. `cb` is not the correct type. You will likely need to somehow wrap `cb` in a function that implements the correct type expected by `aws_iot_mqtt_subscribe`.

Recall that this is the callback type accepted by `aws_iot_mqtt_subscribe`:

```c
typedef void(* pApplicationHandler_t) (AWS_IoT_Client *pClient, char *pTopicName, uint16_t topicNameLen, IoT_Publish_Message_Params *pParams, void *pClientData)
```

And this is the callback type accepted by `nvds_msgapi_subscribe`:

```c
typedef void(* nvds_msgapi_subscribe_request_cb_t)(NvDsMsgApiErrorType flag, void *msg, int msg_len, char *topic, void *user_ptr)
```

So we'll need to create a function of type `*pApplicationHandler_t` that calls `cb`.

Something **like** this (not this exactly) would work:

```c
void nvds_msgapi_subscribe_cb_wrapped(AWS_IoT_Client *pClient, char *pTopicName, uint16_t topicNameLen, IoT_Publish_Message_Params *pParams, void *pClientData)
{
    cb(NVDS_MSGAPI_OK, pParams->payload, pParams->payloadLen, pTopicName, pClientData);
}
```

The trick here is determining how to create this wrapped callback. You will likely need to have a static variable that stores a pointer to such a function, then you can set the value of this static variable when the subscribe method is called.

lummish commented on aws_protocol_adaptor/device_client/aws_nvmsgbroker.c:
```diff
@@ -243,41 +243,32 @@ NvDsMsgApiErrorType nvds_msgapi_send_async(NvDsMsgApiHandle h_ptr, char *topic,
 	{
-		IOT_ERROR("Unable to subscribe, error: %d\n", rc);
-		return NVDS_MSGAPI_ERR;
+		rc = aws_iot_mqtt_subscribe(client, topics[i], strlen(topics[i]), QOS0, cb, user_ctx);
```
This will not compile. `cb` is not the correct type. You will likely need to somehow wrap `cb` in a function that implements the correct type expected by `aws_iot_mqtt_subscribe`.

Recall that this is the callback type accepted by `aws_iot_mqtt_subscribe`:

```c
typedef void(* pApplicationHandler_t) (AWS_IoT_Client *pClient, char *pTopicName, uint16_t topicNameLen, IoT_Publish_Message_Params *pParams, void *pClientData)
```

And this is the callback type accepted by `nvds_msgapi_subscribe`:

```c
typedef void(* nvds_msgapi_subscribe_request_cb_t)(NvDsMsgApiErrorType flag, void *msg, int msg_len, char *topic, void *user_ptr)
```

So we'll need to create a function of type `*pApplicationHandler_t` that calls `cb`.

Something **like** this (not this exactly) would work:

```c
void nvds_msgapi_subscribe_cb_wrapped(AWS_IoT_Client *pClient, char *pTopicName, uint16_t topicNameLen, IoT_Publish_Message_Params *pParams, void *pClientData)
{
    cb(NVDS_MSGAPI_OK, pParams->payload, pParams->payloadLen, pTopicName, pClientData);
}
```

The trick here is determining how to create this wrapped callback. You will likely need to have a static variable that stores a pointer to such a function, then you can set the value of this static variable when the subscribe method is called for the first time.

Shubhamturakhia replied:
I was actually looking for way of wrapping the callback, thanks for the above explanation

lummish replied:
@Shubhamturakhia Here's an idea:

Define a static variable `nvds_cb` that we will use to store the callback provided by a user:

```c
static nvds_msgapi_subscribe_request_cb_t nvds_cb;
```

Define a function that implements the type of callbacks expected by the mqtt subscribe function:

```c
void nvds_cb_wrapped(AWS_IoT_Client *pClient, char *pTopicName, uint16_t topicNameLen, IoT_Publish_Message_Params *pParams, void *pClientData)
{
    nvds_cb(NVDS_MSGAPI_OK, pParams->payload, pParams->payloadLen, pTopicName, pClientData);
}
```

In your implementation of nvds_msgapi_subscribe, assign the callback supplied as an argument to the `nvds_cb` variable:

```c
nvds_cb = cb;
```

lummish reviewed (changes requested):
You need to make sure that you're actually using the correct types in your implementation. Don't just try to implement something intuitively—read the documentation.

This still won't even compile. Please see comment for suggestions.

Shubhamturakhia committed:
Add callback wrapper

Shubhamturakhia committed:
Added unsubscribe feature

lummish commented on aws_protocol_adaptor/device_client/aws_nvmsgbroker.c:
```diff
@@ -33,6 +33,7 @@
 #include "aws_config_parser.h"
 #include "nvds_msgapi.h"
 #include "aws_nvmsgbroker.h"
+#define MAX_SUBSCRIPTIONS 50
```
I think the `aws_iot_config.h` file has a bunch of similar definitions in it. This probably belongs there.

lummish commented on aws_protocol_adaptor/device_client/aws_nvmsgbroker.c:
```diff
@@ -41,12 +42,16 @@ static GMutex thread_mutex;
 static GQueue *work_queue;
 static struct timespec last_send_time_stamp;   // this is to make sure we send or yield frequent enough so we do not get disconnected.
 static nvds_msgapi_connect_cb_t disconnect_cb; // disconnect handler provided by connect thread
+static nvds_msgapi_subscribe_request_cb_t nvds_cb;
+static char *subscribed_topics[MAX_SUBSCRIPTIONS];
+static size_t num_subscriptions = 0;
```
Please add comments explaining these variables.

lummish commented on aws_protocol_adaptor/device_client/aws_nvmsgbroker.c:
```diff
@@ -142,11 +147,22 @@ NvDsMsgApiErrorType nvds_msgapi_disconnect(NvDsMsgApiHandle h_ptr)
+		if (SUCCESS != rc)
+		{
+			IOT_ERROR("Unable to unsubscribe, error: %d\n", rc);
+			return NVDS_MSGAPI_ERR;
```
I think in this case it's probably better to simply continue after logging the error, as we want to ensure that the disconnection actually occurs.

lummish reviewed (changes requested):
A few minor changes.

Shubhamturakhia committed:
Changes as per PR review

Shubhamturakhia committed:
Minor change

Shubhamturakhia committed:
Minor change

lummish commented on aws_protocol_adaptor/device_client/aws_nvmsgbroker.c:
```diff
@@ -33,18 +33,17 @@
-static nvds_msgapi_subscribe_request_cb_t nvds_cb;
-static char *subscribed_topics[MAX_SUBSCRIPTIONS];
-static size_t num_subscriptions = 0;
+static struct timespec last_send_time_stamp;	   // this is to make sure we send or yield frequent enough so we do not get disconnected.
+static nvds_msgapi_connect_cb_t disconnect_cb;	   // disconnect handler provided by connect thread
+static nvds_msgapi_subscribe_request_cb_t nvds_cb; // msgapi subscribe callback handler
+static char *subscribed_topics[MAX_SUBSCRIPTIONS]; // to store the subscribed topics in order to be used during unsubscribe operation
+static size_t num_subscriptions = 0;			   // initializing the number of topics subscribed variable
```
Wonky spacing here.

lummish reviewed (changes requested):
Minor formatting changes required

Shubhamturakhia committed:
Fix formatting

lummish commented on aws_protocol_adaptor/device_client/aws_nvmsgbroker.c:
```diff
@@ -39,11 +39,11 @@ NvDsMsgApiErrorType (*nvds_msgapi_send_ptr)(NvDsMsgApiHandle conn, char *topic,
 static nvds_msgapi_subscribe_request_cb_t nvds_cb; // msgapi subscribe callback handler
 static char *subscribed_topics[MAX_SUBSCRIPTIONS]; // to store the subscribed topics in order to be used during unsubscribe operation
-static size_t num_subscriptions = 0;			   // initializing the number of topics subscribed variable
+static size_t num_subscriptions = 0; // initializing the number of topics subscribed variable
```
You don't need to tell us that you're initializing a variable—that's patently obvious. What we want to know is what the variable means.

```suggestion
static size_t num_subscriptions = 0; // number of registered subscriptions
```

lummish reviewed (changes requested):
One change request

Shubhamturakhia committed:
Update aws_protocol_adaptor/device_client/aws_nvmsgbroker.c

Co-authored-by: Harris Lummis <harris@automotus.co>

lummish reviewed (dismissed):
LGTM

Shubhamturakhia committed:
Add msgapi connection signature mwthod

Shubhamturakhia committed:
Merge branch 'DS-5.0-subscription-implementation' of https://github.com/automotus/aws-iot-core-integration-with-nvidia-deepstream into DS-5.0-subscription-implementation

Shubhamturakhia committed:
added connection signature method

Shubhamturakhia committed:
Minor changes

Shubhamturakhia committed:
Modified utilities for conn signature

Shubhamturakhia committed:
Formatting and minor refactoring

lummish commented on aws_protocol_adaptor/device_client/aws_nvmsgbroker.h:
```diff
@@ -41,4 +41,6 @@ NvDsMsgApiHandle nvds_msgapi_connect(char *connection_str, nvds_msgapi_connect_c
 NvDsMsgApiErrorType nvds_msgapi_send_async(NvDsMsgApiHandle h_ptr, char *topic, const uint8_t *payload, size_t nbuf, nvds_msgapi_send_cb_t send_callback, void *user_ptr);
+NvDsMsgApiErrorType nvds_msgapi_subscribe(NvDsMsgApiHandle h_ptr, char **topics, int num_topics, nvds_msgapi_subscribe_request_cb_t cb, void *user_ctx);
 void nvds_msgapi_do_work(NvDsMsgApiHandle h_ptr);
+NvDsMsgApiErrorType nvds_msgapi_connection_signature(char *broker_str, char *cfg, char *output_str, int max_len);
```
I don't see `get_version` or `get_protocol_name` defined in the header file

Shubhamturakhia replied:
The functions have constants defined in the iot_config file which directly give out the string values when called.

lummish replied:
Not what I'm saying. These functions are not exposed as part of the library's public API because they're not declared in this header file.

lummish commented on aws_protocol_adaptor/device_client/aws_nvmsgbroker.c:
```diff
@@ -353,3 +355,80 @@ void nvds_msgapi_do_work(NvDsMsgApiHandle h_ptr)
+
+char *nvds_msgapi_getversion()
+{
+	return (char *)NVDS_MSGAPI_VERSION;
```
`NVDS_MSGAPI_VERSION` is not a string literal, it's a float.

Shubhamturakhia replied:
According to the documentation, this function returns a string that identifies the nvds_msgapi version, so I implemented according to that, please let me know if you need any changes

lummish replied:
That's all fine—the issue was that the macro was of the incorrect type. I see it's now fixed.

lummish commented on aws_protocol_adaptor/device_client/aws_nvmsgbroker.c:
```diff
@@ -353,3 +355,80 @@ void nvds_msgapi_do_work(NvDsMsgApiHandle h_ptr)
+
+char *nvds_msgapi_get_protocol_name()
+{
+	return (char *)NVDS_MSGAPI_PROTOCOL;
```
You shouldn't need to cast this

Shubhamturakhia replied:
Just clarifying on this, what should I return instead of this?

lummish replied:
That's what you should return, you just don't need to add the explicit cast.

Shubhamturakhia replied:
Does that go for the get_version method as well?

lummish replied:
Yes

lummish commented on aws_protocol_adaptor/device_client/aws_nvmsgbroker.c:
```diff
@@ -353,3 +355,80 @@ void nvds_msgapi_do_work(NvDsMsgApiHandle h_ptr)
+	char conn_str[] = connection_str;
+	int i = 0;
+
+	char *token = strtok(conn_str, ";");
+	char *data[2];
+
+	while (token)
+	{
+		data[i++] = token;
+		token = strtok(NULL, ";");
+	}
```
Don't think this will work properly. The second call to `strtok` shouldn't give you a second token. Instead, try something like this:

```c
int i = 0;
char *data[2];

for (i = 0; i < 2; i++) {
  data[i] = strtok(conn_str, ";");
}

lummish commented on aws_protocol_adaptor/device_client/aws_nvmsgbroker.c:
```diff
@@ -353,3 +355,80 @@ void nvds_msgapi_do_work(NvDsMsgApiHandle h_ptr)
+	return (char *)NVDS_MSGAPI_PROTOCOL;
+}
+
+bool is_valid_connection_str(char *connection_str, char *&burl, char *&bport)
```
What are these types? `char *&`?

lummish replied:
Why are you providing `burl` and `bport`? I thought you were parsing these from the connection string?

lummish commented on aws_protocol_adaptor/device_client/aws_nvmsgbroker.c:
```diff
@@ -353,3 +355,80 @@ void nvds_msgapi_do_work(NvDsMsgApiHandle h_ptr)
+		return NVDS_MSGAPI_ERR;
+	}
+	char *burl = "", *bport = "";
+	if (!is_valid_connection_str(broker_str, burl, bport))
```
Always enclose conditional blocks with curly braces

lummish commented on aws_protocol_adaptor/device_client/aws_nvmsgbroker.c:
```diff
@@ -353,3 +355,80 @@ void nvds_msgapi_do_work(NvDsMsgApiHandle h_ptr)
+		return NVDS_MSGAPI_ERR;
+
+	char *aws_connection_signature = generate_sha256_hash(burl + bport);
+	strcpy(output_str, aws_connection_signature);
```
Why call `strcpy` here instead of using `output_str` to contain the sha256 hash directly?

lummish commented on aws_protocol_adaptor/device_client/aws_nvmsgbroker.c:
```diff
@@ -353,3 +355,80 @@ void nvds_msgapi_do_work(NvDsMsgApiHandle h_ptr)
+	return res;
+}
+
+NvDsMsgApiErrorType nvds_msgapi_connection_signature(char *broker_str, char *cfg, char *output_str, int max_len)
```
`max_len` isn't used anywhere. This suggests that your string operations are unsafe.

lummish commented on aws_protocol_adaptor/device_client/aws_nvmsgbroker.c:
```diff
@@ -353,3 +355,80 @@ void nvds_msgapi_do_work(NvDsMsgApiHandle h_ptr)
+{
+	unsigned char hashval[SHA256_DIGEST_LENGTH];
+	int len = SHA256_DIGEST_LENGTH * 2 + 1;
+	char res[len];
```
`res` is a statically allocated variable, but you're trying to return it. What do you think will happen as a result?

lummish replied:
To remedy the problem, you should supply the output string as an argument to the function. Note: it will need to have space for at least 65 characters (which you'll need to add a check for).

Shubhamturakhia replied:
The max_len is actually the length of the output_str and I have added a check in the connection signature to check if has at least 65 characters

lummish replied:
Sounds good, but you still need to resolve the issue of returning a statically allocated variable. res is created on the stack, not the heap, so its lifetime is only that of the enclosing scope. The value of res will be invalidated once this function returns.

lummish commented on aws_protocol_adaptor/device_client/aws_nvmsgbroker.c:
```diff
@@ -353,3 +355,80 @@ void nvds_msgapi_do_work(NvDsMsgApiHandle h_ptr)
+	SHA256_Final(hashval, &sha256);
+	for (int i = 0; i < SHA256_DIGEST_LENGTH; i++)
+	{
+		res + (i * 2), "%02x", hashval[i];
```
Was this supposed to be a function call?

lummish commented on aws_protocol_adaptor/device_client/aws_nvmsgbroker.c:
```diff
@@ -353,3 +355,80 @@ void nvds_msgapi_do_work(NvDsMsgApiHandle h_ptr)
+	if (!is_valid_connection_str(broker_str, burl, bport))
+		return NVDS_MSGAPI_ERR;
+
+	char *aws_connection_signature = generate_sha256_hash(burl + bport);
```
Why do you need to split the connection string in the first place? Couldn't you just supply the connection string on its own? Also, this is not how you concatenate strings in c.

lummish reviewed (changes requested):
Several head-scratchers

Shubhamturakhia committed:
Changes as per PR review

lummish commented on aws_protocol_adaptor/device_client/aws_nvmsgbroker.c:
```diff
@@ -300,3 +355,84 @@ void nvds_msgapi_do_work(NvDsMsgApiHandle h_ptr)
+
+NvDsMsgApiErrorType nvds_msgapi_connection_signature(char *broker_str, char *cfg, char *output_str, int max_len)
+{
+	strcpy(output_str, "");
```
Why?

Shubhamturakhia replied:
As per the documentation, the output_str would give the required hash code if successful or return empty string if unsuccessful so just initializing it as empty string and then updating it with code if successful, please let me know if you require any changes

lummish replied:
@Shubhamturakhia Got it. Add a comment to explain this line but the code is good.

lummish commented on aws_protocol_adaptor/device_client/aws_nvmsgbroker.c:
```diff
@@ -300,3 +355,84 @@ void nvds_msgapi_do_work(NvDsMsgApiHandle h_ptr)
+		IOT_ERROR("nvds_msgapi_connection_signature: broker_str or cfg path cant be NULL\n");
+		return NVDS_MSGAPI_ERR;
+	}
+	if(max_len < 2 * SHA256_DIGEST_LENGTH + 1) {
```
Can you explain?

Also, please add a space after "if"

Shubhamturakhia replied:
The Sha_digest_length is defined as 32 in the openssl/sha.h so 2*32+1= 65, so if the output_str is less than 65 chars it would return an error

lummish replied:
@Shubhamturakhia Consider defining a macro instead. This is hard to read.

lummish commented on aws_protocol_adaptor/device_client/aws_nvmsgbroker.c:
```diff
@@ -300,3 +355,84 @@ void nvds_msgapi_do_work(NvDsMsgApiHandle h_ptr)
+	if(max_len < 2 * SHA256_DIGEST_LENGTH + 1) {
+        IOT_ERROR("nvds_msgapi_connection_signature: insufficient output string length. Atleast %d needed", 2 * SHA256_DIGEST_LENGTH + 1);
+        return NVDS_MSGAPI_ERR;
+    }
```
Inconsistent indentation.

lummish reviewed (changes requested):
Left comments and added a few requested changes.

Shubhamturakhia committed:
Modifications for Hash code

lummish commented on aws_protocol_adaptor/device_client/aws_nvmsgbroker.c:
```diff
@@ -401,7 +401,7 @@ bool is_valid_connection_str(char *connection_str)
 }
 
-char *generate_sha256_hash(char *str)
+char *generate_sha256_hash(char *str, char *output_str)
```
It's typical in C-style APIs for the destination variable to come first. See `sprintf` for an example.

lummish replied:
Also, you're passing in this argument `output_str` but you're not actually using it anywhere.

lummish commented on aws_protocol_adaptor/device_client/aws_nvmsgbroker.c:
```diff
@@ -419,20 +419,22 @@ char *generate_sha256_hash(char *str)
 NvDsMsgApiErrorType nvds_msgapi_connection_signature(char *broker_str, char *cfg, char *output_str, int max_len)
 {
-	strcpy(output_str, "");
+	strcpy(output_str, ""); // Initializing output_str as empty string if successful which is updated later by SHA-256 hash else empty string is returned
```
This comment isn't very helpful. 
```suggestion
	// Value of output_str must be empty string if operation is unsuccessful
	strcpy(output_str, "");
```

lummish commented on aws_protocol_adaptor/device_client/aws_nvmsgbroker.c:
```diff
@@ -419,20 +419,22 @@ char *generate_sha256_hash(char *str)
+	{
+        IOT_ERROR("nvds_msgapi_connection_signature: insufficient output string length. Atleast %d needed", required_output_str_len);
         return NVDS_MSGAPI_ERR;
     }
```
Inconsistent indentation.

lummish reviewed (changes requested):
Several issues

Shubhamturakhia committed:
Changes as per review

Shubhamturakhia committed:
Minor indentation change

Shubhamturakhia committed:
Hash code refactoring

Shubhamturakhia committed:
Add hash code implementation for config file

lummish committed:
Fix Incorrect Return Types

generate_sha256_hash function returned incorrect types upon error.

Refactor logic to separate file reading from connection string hashing
and use integer return type to indicate errors.

Shubhamturakhia committed:
Minor changes

lummish committed:
Resolve Compilation Errors

lummish committed:
Link OpenSSL in Makefile

lummish committed:
Statically Link to libssl and libcrypto

lummish committed:
Reorder SSL Linkages

lummish committed:
Reorder Linkage to SSL Libraries
